src/core_ocean/shared/mpas_ocn_diagnostics.F

```diff
@@ -604,43 +735,523 @@ subroutine ocn_diagnostic_solve(dt, statePool, forcingPool, meshPool, diagnostic
          !$omp end do
          !$omp end parallel
 
-         deallocate(vorticityGradientNormalComponent, &
-                    vorticityGradientTangentialComponent)
+         deallocate(vorticityGradientNormalComponent, &
+                    vorticityGradientTangentialComponent)
+
+      endif
+
+      deallocate(normalizedPlanetaryVorticityVertex, &
+                 normalizedRelativeVorticityVertex)
+
+
+   end subroutine ocn_diagnostic_solve_vorticity!}}}
+
+!***********************************************************************
+!
+!  routine ocn_diagnostic_solve_richardson
+!
+!> \brief   Computes diagnostic variables for gradient richardson
+!> \author  Matt Turner
+!> \date    October 2020
+!> \details
+!>  This routine computes the diagnostic variables for gradient richardson
+!>   number and brunt viasala
+!
+!-----------------------------------------------------------------------
+   subroutine ocn_diagnostic_solve_richardson(displacedDensity, density, zMid, &
+                normalVelocity, edgeAreaFractionOfCell, BruntVaisalaFreqTop, RiTopOfCell)!{{{
+
+      use ocn_mesh   ! debug, mdt :: this should be module-level
+
+      implicit none
+
+      !-----------------------------------------------------------------
+      !
+      ! input variables
+      !
+      !-----------------------------------------------------------------
+
+      real (kind=RKIND), dimension(:,:), intent(in) :: &
+        displacedDensity
+      real (kind=RKIND), dimension(:,:), intent(in) :: &
+        density
+      real (kind=RKIND), dimension(:,:), intent(in) :: &
+        zMid
+      real (kind=RKIND), dimension(:,:), intent(in) :: &
+        normalVelocity
+      real (kind=RKIND), dimension(:,:), intent(in) :: &
+        edgeAreaFractionOfCell
+
+      !-----------------------------------------------------------------
+      !
+      ! output variables
+      !
+      !-----------------------------------------------------------------
+
+      real (kind=RKIND), dimension(:,:), intent(out) :: &
+        BruntVaisalaFreqTop
+      real (kind=RKIND), dimension(:,:), intent(out) :: &
+        RiTopOfCell
+
+      !-----------------------------------------------------------------
+      !
+      ! local variables
+      !
+      !-----------------------------------------------------------------
+
+      real(kind=RKIND) :: coef, shearSquared, delU2, shearMean
+      integer :: iCell, k, i, iEdge
+      integer :: nCells
+      real(kind=RKIND) :: invAreaCell1
+
+      !
+      ! Brunt-Vaisala frequency (this has units of s^{-2})
+      !
+      coef = -gravity / rho_sw
+      !$omp parallel
+      !$omp do schedule(runtime) private(k)
+      do iCell = 1, nCellsAll
+         BruntVaisalaFreqTop(1,iCell) = 0.0_RKIND
+         do k = 2, maxLevelCell(iCell)
+            BruntVaisalaFreqTop(k,iCell) = coef * (displacedDensity(k-1,iCell) - density(k,iCell)) &
```

Comment:
@mattdturner and @philipwjones here is a question: Would it be beneficial to merge loops where possible? Here we have a loop over iCell and k, and there is another one below at line 830. The loops share the `BruntVaisalaFreqTop` variable. As a rule of thumb, should we avoid separate loops when possible? I'm guessing the benefit is reduced memory access, so there is only a speed-up if the loops share variables, so the single loop reduces memory retrievals.

On the flip side, separate loops can be more clear, since each variable, like Brunt-Vaisala here, get their own block of code.

Reply:
This loop is over `nCellsAll`, whereas the loop at line 830 is over `nCellsHalo( 2 )`.  So these 2 loops probably couldn't be merged without some conditionals or other trickery.  

Generally speaking, I try to fuse loops when possible to reduce the overhead associated with more loops.  Loop fission, though, can be beneficial if the number of computations in a loop becomes excessive, leading to register spills that degrade performance.  So unfortunately there isn't a 1 size fits all solution.